src/store_swapout.cc

```diff
@@ -358,20 +357,26 @@ StoreEntry::mayStartSwapOut()
         return false;
     }
 
-    // if we swapped out already, do not start over
-    if (swap_status == SWAPOUT_DONE) {
+    // if we are swapping out or swapped out already, do not start over
+    if (hasDisk() || Store::Root().hasReadableDiskEntry(*this)) {
         debugs(20, 3, "already did");
         swapOutDecision(MemObject::SwapOut::swImpossible);
         return false;
     }
 
-    // if we stared swapping out already, do not start over
+    // if we have just stared swapping out (attachToDisk() has not been
+    // called), do not start over
     if (decision == MemObject::SwapOut::swStarted) {
         debugs(20, 3, "already started");
         swapOutDecision(MemObject::SwapOut::swImpossible);
         return false;
     }
 
+    if (Store::Root().markedForDeletionAndAbandoned(*this)) {
+        debugs(20, 3, "marked for deletion and abandoned");
+        return false;
```

Comment:
Should this be `MemObject::SwapOut::swImpossible`?

Reply:
Yes, it should. I think this bug is a consequence of code duplication throughout this method: swImpossible state is always followed by returning 'false'. I would suggest to fix this with something like:

```
static bool CreateImpossibleDecision(StoreEntry *e, const char *msg)
{
    debugs(20, 3, msg << " " << *e);
    swapOutDecision(MemObject::SwapOut::swImpossible);
    return false;
}
```

and the above block(and other similar checks) would transform into:

```
if (Store::Root().markedForDeletionAndAbandoned(*this))
   return CreateImpossibleDecision(this, "marked for deletion and abandoned");

```

Reply:
Let's not polish this now. I am concerned that either something went wrong in this method (before this PR) or the method itself needs to be redesigned again (outside of this PR) because false always means swImpossible. Addressing that concern is not trivial and is outside this PR scope.

Reply:
Done without polishing at a40bbf6.